File: DailyChallenges/CPP/LeetCode_1466.cpp
class Solution {
public:
    int minReorder(int n, vector<vector<int>>& connections) {

        unordered_map<int, vector<int>> m_st, m_ed;
        for (auto & a: connections){
            m_st[a[0]].push_back(a[1]);
            m_ed[a[1]].push_back(a[0]);
        } 

        int ans = 0;
        queue<int> q; q.push(0);
        vector<int> visited(n, 0); visited[0] = 1;

        while (!q.empty()){
            int sz = q.size();
            for (int i = 0; i<sz; ++i){
                int node = q.front(); q.pop();
                for (int st: m_ed[node]){ q.push(st); visited[st] = 1;}
                for (int ed: m_st[node]){
                    if (!visited[ed]){
                        ans++; q.push(ed); visited[ed] =1;
                    }
                }
            }
        }

        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_1466.scala
object Solution {
    def minReorder(n: Int, connections: Array[Array[Int]]): Int = {
        val m = Array.fill[List[Int]](n)(Nil)
        for (a <- connections) {
            m(a(0)) = a(1) :: m(a(0)) // st -> ed
            m(a(1)) = -a(0) :: m(a(1)) // ed -> -st
        }

        val vt = Array.fill[Boolean](n)(false)

        def dfs(st: Int): Int = {
            var change = 0
            vt(st) = true
            for (ed <- m(st) if !vt(Math.abs(ed))){
                if (ed > 0) change += 1
                change += dfs(Math.abs(ed))
            }
            change
        }

        dfs(0)
    }
}
